cordapp-cpk2/src/main/java/net/corda/plugins/cpb2/CpbTask.java

```diff
@@ -61,18 +65,29 @@ public CpbTask() {
     @NotNull
     public AbstractCopyTask from(@NotNull Object... args) {
         return super.from(args, copySpec ->
-            copySpec.exclude(this::isCPK)
+            copySpec.exclude(this::isValidCPK)
         );
     }
 
-    private boolean isCPK(@NotNull FileTreeElement element) {
+    private boolean isValidCPK(@NotNull FileTreeElement element) {
         if (!element.getName().endsWith(CPK_FILE_SUFFIX)) {
             return false;
         }
 
-        final Path cpkPath = element.getFile().toPath();
+        Path cpkPath = element.getFile().toPath();
         try (JarInputStream cpkStream = new JarInputStream(new BufferedInputStream(Files.newInputStream(cpkPath)))) {
-            final String cpkType = cpkStream.getManifest().getMainAttributes().getValue(CORDA_CPK_TYPE);
+            Manifest manifest = cpkStream.getManifest();
+            String cpkType = manifest.getMainAttributes().getValue(CORDA_CPK_TYPE);
+            String cpkCordappName = manifest.getMainAttributes().getValue(CPK_CORDAPP_NAME);
+            if (cpkCordappName != null) {
+                if (cpkNames.containsKey(cpkCordappName)) {
+                    if (!cpkNames.get(cpkCordappName).equals(manifest)) {
+                        throw new InvalidUserDataException("Two CPKs may not share a cordappCpkName. Error in " + cpkCordappName);
+                    }
+                } else {
+                    cpkNames.put(cpkCordappName, manifest);
+                }
+            }
             return cpkType != null && EXCLUDED_CPK_TYPES.contains(cpkType.toLowerCase());
```

Comment:
(again before this PR)

the body of this function seems a bit misaligned with its name? The initial check (at line 73) does check if it's a CPK but then here we check if its `Corda-CPK-Type` (which I am not sure we currently use) is of of excluded cpk types?

Reply:
It might make more sense if I split it into 3 functions that all get called by `isValidCPK`. Or maybe just 2 if Corda-CPK-Type isn't used anywhere?

Reply:
Yes, I'd not spotted this before but
```
copySpec.exclude(this::isCPK)
```
seems to be excluding based on a predicate.

That would make "hijacking" this method fairly inappropriate for this same check. I'd revert this method to `isCPK` and make this logic it's own thing separate from that process.